lib/puppet/provider/agent_upgrade/windows.rb

```diff
@@ -0,0 +1,85 @@
+
+Puppet::Type.type(:agent_upgrade).provide(:agent_upgrade) do
+  confine :osfamily => /windows/i
+
+  mk_resource_methods
+
+  commands :powershell =>
+             if File.exists?("#{ENV['SYSTEMROOT']}\\sysnative\\WindowsPowershell\\v1.0\\powershell.exe")
+               "#{ENV['SYSTEMROOT']}\\sysnative\\WindowsPowershell\\v1.0\\powershell.exe"
+             elsif File.exists?("#{ENV['SYSTEMROOT']}\\system32\\WindowsPowershell\\v1.0\\powershell.exe")
+               "#{ENV['SYSTEMROOT']}\\system32\\WindowsPowershell\\v1.0\\powershell.exe"
+             else
+               'powershell.exe'
+             end
```

Comment:
Typically you create this so you can `powershell(somearg, arg2, argN)` - in the way that you ended up using it, it may be better to use a variable instead.